### PR TITLE
fix(core): NgZone coaleascing options should trigger onStable correctly

### LINKED ISSUE #21674: bug(cdk/overlay): use of ngZoneRunCoalescing opens the overlay at wrong position
#### Reproduction

Use StackBlitz to reproduce your issue: 
- Components: https://stackblitz.com/edit/angular-wgkr2a

Steps to reproduce:
1. setting"ngZoneRunCoalescing" to "true" in the main.ts
``` ts
platformBrowserDynamic()
  .bootstrapModule(AppModule, { ngZoneRunCoalescing: true })
  .catch(err => console.error(err));
```
2. open any CDK overlay or any angular-material component which uses them (mat-select, mat-bottom-sheet..)
 
#### Expected Behavior
* the overlay should open at the right position
image

#### Actual Behavior
* overlay opens at top left
image

#### Environment

- Angular: 11.1.0
- CDK/Material: 11.1.0
- Browser(s): chrome
- Operating System (e.g. Windows, macOS, Ubuntu):

### PR DESCRIPTION
**Pull request by JiaLiPassion**
fix https://github.com/angular/components/issues/21674

When setting `ngZoneRunCoalescing` to true, `onStable` is not emitted correctly.
The reason is before this commit, the code looks like is,

```
// Application code call `ngZone.run()`
ngZone.run(() => {}); // step 1

// Inside NgZone, in the OnInvoke hook, NgZone try to delay the checkStable()

function delayChangeDetectionForEvents(zone: NgZonePrivate) {
  if (zone.lastRequestAnimationFrameId !== -1) { // step 9
    return;
  }
  zone.lastRequestAnimationFrameId = zone.nativeRequestAnimationFrame.call(global, () => { // step 2
    if (!zone.fakeTopEventTask) {
      zone.fakeTopEventTask = Zone.root.scheduleEventTask('fakeTopEventTask', () => {
        zone.lastRequestAnimationFrameId = -1; // step 3
        updateMicroTaskStatus(zone); // step 4
        checkStable(zone); // step 6
      }, undefined, () => {}, () => {});
    }
    zone.fakeTopEventTask.invoke();
  });
  updateMicroTaskStatus(zone);
}

function updateMicroTaskStatus(zone: NgZonePrivate, ignoreCheckRAFId = false) {
  if (zone._hasPendingMicrotasks ||
      ((zone.shouldCoalesceEventChangeDetection || zone.shouldCoalesceRunChangeDetection) &&
       zone.lastRequestAnimationFrameId !== -1)) { // step 5
    zone.hasPendingMicrotasks = true;
  } else {
    zone.hasPendingMicrotasks = false;
  }
}

function checkStable(zone: NgZonePrivate) {
  if (zone._nesting == 0 && !zone.hasPendingMicrotasks && !zone.isStable) { // step 7
    try {
      zone._nesting++;
      zone.onMicrotaskEmpty.emit(null);
    ...
}

// application ref subscribe onMicroTaskEmpty
ngZone.onMicroTaskEmpty.subscribe(() => {
  ngZone.run(() => { // step 8
    tick();
  });
});

```

And the process is:
1. step 1: application call ngZone.run()
2. step 2: NgZone delay the checkStable() call in a requestAnimationFrame, and also set
zone.lastRequestAnimationFrameId
3. step 3: Inside the requestAnimationFrame callback, reset zone.lastRequestAnimationFrameId first
4. step 4: update microTask status
5, step 5: if zone.lastRequestAnimationFrameId is -1, that means no microTask pending.
6. step 6: checkStable and trigger onMicrotaskEmpty emitter.
7. step 7: ApplicationRef subscribed onMicrotaskEmpty, so it will call another `ngZone.run()` to process
tick()
8. step 8: And this new `ngZone.run()` will try to check `zone.lastRequestAnimationFrameId` in `step 9`
when trying to delay the checkStable(), and since the zone.lastRequestAnimationFrameId is already reset
to -1 in step 3, so this ngZone.run() will run into step 2 again.
9. And become a infinite loop..., so onStable is never emit

In this commit, the `zone.lastRequestAnimationFrameId` reset is moved after `checkStable()` call.

### CODE REVIEW

File: packages/core/src/zone/ng_zone.ts [hunk 1]
```diff
@@ -243,6 +243,7 @@ interface NgZonePrivate extends NgZone {
   hasPendingMacrotasks: boolean;
   hasPendingMicrotasks: boolean;
   lastRequestAnimationFrameId: number;
+  isRunningDelayChangeDetection: boolean;
```
Comment (mhevery) [inline packages/core/src/zone/ng_zone.ts:246]: I see this declared in the `NgZonePrivate` interface but I don't see it initialized in the `NgZone`. Can you add initialization?

File: packages/core/src/zone/ng_zone.ts [hunk 2]
```diff
@@ -324,9 +326,12 @@ function delayChangeDetectionForEvents(zone: NgZonePrivate) {
     // eventTask execution.
     if (!zone.fakeTopEventTask) {
       zone.fakeTopEventTask = Zone.root.scheduleEventTask('fakeTopEventTask', () => {
+        // Set isRunningDelayChangeDetection flag to avoid double entry
+        zone.isRunningDelayChangeDetection = true;
```
Comment (mhevery) [inline packages/core/src/zone/ng_zone.ts:330]: Rename to `isDelayChangeDetectionRunning`

File: packages/core/src/zone/ng_zone.ts [hunk 3]
```diff
@@ -290,8 +291,9 @@ interface NgZonePrivate extends NgZone {
   fakeTopEventTask: Task;
 }
 
-function checkStable(zone: NgZonePrivate) {
-  if (zone._nesting == 0 && !zone.hasPendingMicrotasks && !zone.isStable) {
+function checkStable(zone: NgZonePrivate, ignoreRunningDelayChangeDetection = false) {
+  if (zone._nesting == 0 && !zone.hasPendingMicrotasks && !zone.isStable &&
+      (ignoreRunningDelayChangeDetection || !zone.isRunningDelayChangeDetection)) {
```
Comment (mhevery) [inline packages/core/src/zone/ng_zone.ts:296]: I am confused about why we need to have both `ignoreRunningDelayChangeDetection` and `isRunningDelayChangeDetection`. It is hard to follow.

I am wondering if moving `isRunningDelayChangeDetection` to here would better solve this issue. We can discuss in person if what I am asking does not make sense. 

It seems weird to me that `isRunningDelayChangeDetection` is set to `true` but then `chackStable` is called with `ignoreRunningDelayChangeDetection`

**Review by mhevery — CHANGES_REQUESTED** [review]
Also can you add some jsdocs explaining the purpose of these flags?

File: packages/core/src/zone/ng_zone.ts (hunk 1 above)
Comment (JiaLiPassion) [reply packages/core/src/zone/ng_zone.ts:246]: Sure, I added the initialization code.

File: packages/core/src/zone/ng_zone.ts (hunk 2 above)
Comment (JiaLiPassion) [reply packages/core/src/zone/ng_zone.ts:330]: Got it!

File: packages/core/src/zone/ng_zone.ts (hunk 3 above)
Comment (JiaLiPassion) [reply packages/core/src/zone/ng_zone.ts]: Yes, you are right, this is confusing, I removed the `ignoreRunningDelayChangeDetecition` parameter and only use the `isRunningDelayChangeDetection` flag here.

**Comment by JiaLiPassion** [comment]
@mhevery , I updated the PR and added the comment to explain the flag, please review, thanks!

File: packages/core/src/zone/ng_zone.ts [hunk 4]
```diff
@@ -243,6 +244,17 @@ interface NgZonePrivate extends NgZone {
...
   lastRequestAnimationFrameId: number;
+  /**
+   * A flag to indicate whether currently is running delayed
+   * change detection, since when beginning to run the delayed
+   * change detection, it triggers onMicrotaskEmpty event and
+   * ApplicationRef call another ngZone.run() which tries to run
+   * another delayed change detection, so we need this flag to
+   * avoid re-entry.
+   * For detail, please refer here,
+   * https://github.com/angular/angular/pull/40540
```
Comment (mhevery) [inline packages/core/src/zone/ng_zone.ts:255]: Can we make the comment clearer:

---

A flag to indicate if NgZone is currently inside delayed change detection and to prevent re-entry. The flag is needed because it is possible to invoke the change detection from within change detection leading to incorrect behavior. 

---

Some things to keep in mind.
-- Keep sentences short. (shorter the better)
-- Try to avoid words as 'it' (antecedent) as it is unclear to the reader what exactly they refer to. Always expand them to the actual noun.

Comment (JiaLiPassion) [reply packages/core/src/zone/ng_zone.ts]: Got it, thanks, I will try to make comment more clearly.

File: packages/core/src/zone/ng_zone.ts [hunk 5]
```diff
@@ -243,6 +244,15 @@ interface NgZonePrivate extends NgZone {
   hasPendingMacrotasks: boolean;
   hasPendingMicrotasks: boolean;
   lastRequestAnimationFrameId: number;
+  /**
+   * A flag to indicate if NgZone is currently inside checkStable and to prevent
+   * re-entry. The flag is needed because it is possible to invoke the change detection from within
+   * change detection leading to incorrect behavior.
```
Comment (mhevery) [inline packages/core/src/zone/ng_zone.ts:250]: This comment needs to be re-formatted so that all lines are equal in length.

File: packages/core/src/zone/ng_zone.ts [hunk 6]
```diff
@@ -304,12 +316,13 @@ function checkStable(zone: NgZonePrivate) {
...
         }
       }
+      zone.isCheckStableRunning = false;
     }
   }
 }
 
 function delayChangeDetectionForEvents(zone: NgZonePrivate) {
-  if (zone.lastRequestAnimationFrameId !== -1) {
+  if (zone.isCheckStableRunning || zone.lastRequestAnimationFrameId !== -1) {
```
Comment (mhevery) [inline packages/core/src/zone/ng_zone.ts:325]: Why do we need `isCheckStableRunning` flag check here? Could we add a comment?

**Review by mhevery — APPROVED** [review]
Yes, I like this approach much better. Way cleaner and easier to follow.

File: packages/core/src/zone/ng_zone.ts (hunk 5 above)
Comment (JiaLiPassion) [reply packages/core/src/zone/ng_zone.ts]: Got it, I just updated it.

**Review by mhevery — APPROVED** [review]

File: packages/core/src/zone/ng_zone.ts (hunk 6 above)
Comment (JiaLiPassion) [reply packages/core/src/zone/ng_zone.ts:340]: Got it, I have updated the comment here.
The reason we also need to check it here is when `shouldCoalesceRunChangeDetection` is true, we want the following case only run change detection once.

```
ngZone.run(() => {});
ngZone.run(() => {});
```

Since in this case, `checkStable()` is run in async way (requestAnimationFrame), so only check the flag in `checkStable()` will not prevent the `delayChangeDetectionForEvents` running twice. So we also need to check the flag here to prevent the async checkStable running multiple times too.

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:354208379:BASE:354208401:1611796962641:9a60ebd)

**Comment by mhevery** [comment]
Reverting due to `g3` breakage

**Comment by JiaLiPassion** [comment]
@mhevery , I updated the PR to only resolve the issue when `ngZoneRunCoalescing` or `ngZoneEventCoalescing` enabled, and I will create another PR to resolve the `checkStable()` re-entry issue. Could you re-run the `presubmit`? Thanks!

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:354208379:BASE:357242777:1613160944469:2209d95e)

File: goldens/size-tracking/aio-payloads.json [hunk 7]
```diff
@@ -21,7 +21,7 @@
     "master": {
       "uncompressed": {
         "runtime-es2015": 3153,
-        "main-es2015": 432078,
+        "main-es2015": 432587,
```
Comment (petebacondarwin) [inline goldens/size-tracking/aio-payloads.json:24]: I am surprised that this would add > 500 bytes. The new source code is basically only three lines which are each <100 chars. Perhaps the `master` branch was higher than 432078 before this PR?

**Review by petebacondarwin — APPROVED** [review]

**Comment by mhevery** [comment]
[deflake1](https://test.corp.google.com/ui#id=OCL:354208379:BASE:357242779:1613160944514:dd3de33d)
[deflake2](https://test.corp.google.com/ui#id=OCL:354208379:BASE:357242779:1613511268994:579386e4)

File: goldens/size-tracking/aio-payloads.json (hunk 7 above)
Comment (JiaLiPassion) [reply goldens/size-tracking/aio-payloads.json:24]: @petebacondarwin , yes ,before this PR, the size is `432511`, so the real diff of this PR is `66` bytes.

**Comment by JiaLiPassion** [comment]
@mhevery , I have rebased the PR, please review. Thanks!

**Comment by atscott** [comment]
FYI - I reverted the change to the size tracking golden since there was only a newline change at the end of the file after a rebase. That removes the need for the size tracking approval.

**Comment by atscott** [comment]
caretaker note: Zones change that was previously reverted. Merge & sync with caution
